types/src/network_address.rs

```diff
@@ -0,0 +1,645 @@
+    UnexpectedEnd,
+
+    #[error("error parsing ip4/ip6 address: {0}")]
+    ParseAddrError(#[from] net::AddrParseError),
+
+    #[error("error parsing integer: {0}")]
+    ParseIntError(#[from] num::ParseIntError),
+
+    #[error("error parsing x25519 public key: {0}")]
+    ParseX25519PubkeyError(#[from] CryptoMaterialError),
+
+    #[error("network address cannot be empty")]
+    EmptyProtocolString,
+
+    #[error("protocol string must start with '/'")]
+    InvalidProtocolString,
+
+    #[error("dns name cannot be empty")]
+    EmptyDnsNameString,
+
+    #[error("dns name cannot contain '/' characters")]
+    InvalidDnsNameCharacter,
+
+    #[error("dns name is too long: len: {0} bytes, max len: 255 bytes")]
+    DnsNameTooLong(usize),
+}
+
+#[derive(Error, Debug)]
+#[error("network address cannot be empty")]
+pub struct EmptyError;
+
+///////////////////////
+// RawNetworkAddress //
+///////////////////////
+
+impl RawNetworkAddress {
+    pub fn new(bytes: Vec<u8>) -> Self {
+        Self(bytes)
+    }
+
+    pub fn empty() -> Self {
+        Self(Vec::new())
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
+
+impl Into<Vec<u8>> for RawNetworkAddress {
+    fn into(self) -> Vec<u8> {
+        self.0
+    }
+}
+
+impl AsRef<[u8]> for RawNetworkAddress {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl TryFrom<&NetworkAddress> for RawNetworkAddress {
+    type Error = lcs::Error;
+
+    fn try_from(value: &NetworkAddress) -> Result<Self, lcs::Error> {
+        let bytes = lcs::to_bytes(value)?;
+        Ok(RawNetworkAddress::new(bytes))
+    }
+}
+
+////////////////////
+// NetworkAddress //
+////////////////////
+
+impl NetworkAddress {
+    fn new(protocols: Vec<Protocol>) -> Self {
+        Self(protocols)
+    }
+}
+
+impl FromStr for NetworkAddress {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.is_empty() {
+            return Err(ParseError::EmptyProtocolString);
+        }
+
+        let mut protocols = Vec::new();
+        let mut parts = s.split('/');
+
+        if parts.next() != Some("") {
+            return Err(ParseError::InvalidProtocolString);
+        }
+
+        while let Some(protocol_type) = parts.next() {
+            let arg = parts.next().ok_or(ParseError::UnexpectedEnd)?;
+            let protocol = match protocol_type {
+                "ip4" => Protocol::Ip4(arg.parse()?),
+                "ip6" => Protocol::Ip6(arg.parse()?),
+                "dns" => Protocol::Dns(arg.parse()?),
+                "dns4" => Protocol::Dns4(arg.parse()?),
+                "dns6" => Protocol::Dns6(arg.parse()?),
+                "tcp" => Protocol::Tcp(arg.parse()?),
+                "memory" => Protocol::Memory(arg.parse()?),
+                "ln-noise-ik" => Protocol::NoiseIk(x25519::PublicKey::from_encoded_string(arg)?),
```

Comment:
Is ln-noise different than just noise?

Reply:
with the timestamp in the handshake prologue for replay attack prevention (still todo), our handshake will be libra specific. i figured i would signal "libra-specific protocol" by prepending "ln-"; of course, since this is a non-standard address format anyway, then maybe that's not necessary since "libra-specific" is already implied?